### PR TITLE
Extra: widgets: add a new widget animation image

### LINKED ISSUE #2090: How to support animation format(e.g. gif) through lv_img_decoder_t?
### Introduce the problem

lv_img_decoder_t doesn't have the fields or callback to provide the mutiple frame data, how to write a portable widget to show the animation?

### Examples and cases

Improve img widget so show the animation.

### Suggested solution

Add animation info(e.g. frame_rate, frame_number, loop) to lv_img_header_t or lv_img_decoder_dsc_t. Add new callback(switch the frame) in lv_img_decoder_t.

### PR DESCRIPTION
**Pull request by ZhaoQiang-b45475**
This widget is a combination of img and animation,
could achieve animation effects by
constantly switching a series imgs.

Signed-off-by: Qiang Zhao <qiang.zhao@nxp.com>
Signed-off-by: Hui Song <hui.song_1@nxp.com>
Signed-off-by: Xiaolin He <xiaolin.he@nxp.com>

### Description of the feature or fix

example for use:
LV_IMG_DECLARE(anim_ezgif_frame_001)
LV_IMG_DECLARE(anim_ezgif_frame_002)
LV_IMG_DECLARE(anim_ezgif_frame_003)
LV_IMG_DECLARE(anim_ezgif_frame_004)
LV_IMG_DECLARE(anim_ezgif_frame_005)
LV_IMG_DECLARE(anim_ezgif_frame_006)
LV_IMG_DECLARE(anim_ezgif_frame_007)
LV_IMG_DECLARE(anim_ezgif_frame_008)
LV_IMG_DECLARE(anim_ezgif_frame_009)
LV_IMG_DECLARE(anim_ezgif_frame_010)
LV_IMG_DECLARE(anim_ezgif_frame_011)
LV_IMG_DECLARE(anim_ezgif_frame_012)
LV_IMG_DECLARE(anim_ezgif_frame_013)
LV_IMG_DECLARE(anim_ezgif_frame_014)
LV_IMG_DECLARE(anim_ezgif_frame_015)
LV_IMG_DECLARE(anim_ezgif_frame_016)
LV_IMG_DECLARE(anim_ezgif_frame_017)
LV_IMG_DECLARE(anim_ezgif_frame_018)
LV_IMG_DECLARE(anim_ezgif_frame_019)
LV_IMG_DECLARE(anim_ezgif_frame_020)
LV_IMG_DECLARE(anim_ezgif_frame_021)
LV_IMG_DECLARE(anim_ezgif_frame_022)
LV_IMG_DECLARE(anim_ezgif_frame_023)
LV_IMG_DECLARE(anim_ezgif_frame_024)
LV_IMG_DECLARE(anim_ezgif_frame_025)
LV_IMG_DECLARE(anim_ezgif_frame_026)
LV_IMG_DECLARE(anim_ezgif_frame_027)
LV_IMG_DECLARE(anim_ezgif_frame_028)
LV_IMG_DECLARE(anim_ezgif_frame_029)
LV_IMG_DECLARE(anim_ezgif_frame_030)
LV_IMG_DECLARE(anim_ezgif_frame_031)

static const lv_img_dsc_t*   screen_animimg0_imgs[31] = {
		 &anim_ezgif_frame_001,
		 &anim_ezgif_frame_002,
		 &anim_ezgif_frame_003,
		 &anim_ezgif_frame_004,
		 &anim_ezgif_frame_005,
		 &anim_ezgif_frame_006,
		 &anim_ezgif_frame_007,
		 &anim_ezgif_frame_008,
		 &anim_ezgif_frame_009,
		 &anim_ezgif_frame_010,
		 &anim_ezgif_frame_011,
		 &anim_ezgif_frame_012,
		 &anim_ezgif_frame_013,
		 &anim_ezgif_frame_014,
		 &anim_ezgif_frame_015,
		 &anim_ezgif_frame_016,
		 &anim_ezgif_frame_017,
		 &anim_ezgif_frame_018,
		 &anim_ezgif_frame_019,
		 &anim_ezgif_frame_020,
		 &anim_ezgif_frame_021,
		 &anim_ezgif_frame_022,
		 &anim_ezgif_frame_023,
		 &anim_ezgif_frame_024,
		 &anim_ezgif_frame_025,
		 &anim_ezgif_frame_026,
		 &anim_ezgif_frame_027,
		 &anim_ezgif_frame_028,
		 &anim_ezgif_frame_029,
		 &anim_ezgif_frame_030,
		 &anim_ezgif_frame_031 
 	}; 

void lv_demo(void){
	lv_obj_t * animimg0 = lv_animimg_create(lv_scr_act());
	lv_obj_set_pos(animimg0, 172, 83);
	lv_obj_set_size(animimg0, 100, 100);
	lv_animimg_set_src(animimg0, (lv_img_dsc_t**) screen_animimg0_imgs, 31);
	lv_animimg_set_duration(animimg0, 930);
	lv_animimg_set_repeat_count(animimg0, 3000);
	lv_animimg_start(animimg0);
}

### Checkpoints
- [ ] Follow the [styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [ ] Update CHANGELOG.md
- [ ] Update the documentation

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
@ZhaoQiang-b45475 nice addition!
But the method to specify the image source is too boring and require the user to convert one animation file to the sequence frame.
How about we extend the decoder interface to support the animation widget? We can even provide a decoder which accept the array of image(e.g. screen_animimg0_imgs) and return them in sequence like the current usage.
But it's open the possibility to support more animaiton file(gif, apng, lottiefiles) in the future without modifing the widgets.
Here has more dissussion: https://github.com/lvgl/lvgl/issues/2090

**Comment by ZhaoQiang-b45475** [comment]
The build failed, but it seems not to be caused by my patch, right?

**Comment by xiaoxiang781216** [comment]
> The build failed, but it seems not to be caused by my patch, right?

Please ignore Micropython failure. Micropython doesn't support v8 yet.

**Comment by kisvegabor** [comment]
@xiaoxiang781216 
I think this widget could be useful because of its simplicity. No extra decoder, no special file format, just simple images.

@ZhaoQiang-b45475 
I have some comments on API, mainly to follow LVGL's API conventions

```c
//We just removed the "copy" parameter for v8
lv_obj_t * animimg0 = lv_animimg_create(lv_scr_act(), NULL);

//The number of images should be set with the source to be sure they are consistent
// lv_animimg_set_src(animimg0, (lv_img_dsc_t**) screen_animimg0_imgs, 31);
lv_set_anim_img_sources(animimg0, (lv_img_dsc_t**) screen_animimg0_imgs);

//lv_animimg_set_duration(animimg0, 930)
lv_set_anim_img_duration(animimg0, 930);

//What's the difference with lv_set_anim_img_duration?
lv_set_anim_img_playback_time(animimg0, 930);

//Why we need the word "startup" here?
//lv_animimg_set_repeat_count(animimg0, 3000)
lv_set_anim_img_startup_repeat_count(animimg0, 3000);

//lv_animimg_start(animimg0)
lv_anim_img_startup(animimg0);
```

Please move the files into `extra/widgets` folder similarly to the other widgets there.

**Comment by ZhaoQiang-b45475** [comment]
Hi @kisvegabor ,
Updated, please review.

**Comment by kisvegabor** [comment]
Looks good, thank you!.
One more thing remained: please add an example to the `lvgl/example/widgets/animimg` folder with 2-3 small images.

**Comment by ZhaoQiang-b45475** [comment]
Hi Kisvegabor,
Ok, I have added the example.

**Comment by ZhaoQiang-b45475** [comment]
Hi,
**The check error are caused by image c file.**
Error: /home/runner/work/lvgl/lvgl/tests/../../lvgl/examples/assets/anim002.c:20:79: error: ISO C forbids empty initializer braces [-Wpedantic]
   20 | const LV_ATTRIBUTE_MEM_ALIGN LV_ATTRIBUTE_IMG_ANIM002 uint8_t anim002_map[] = {
      |                                                                               ^
Error: /home/runner/work/lvgl/lvgl/tests/../../lvgl/examples/assets/anim002.c:20:63: error: zero or negative size array ‘anim002_map’
   20 | const LV_ATTRIBUTE_MEM_ALIGN LV_ATTRIBUTE_IMG_ANIM002 uint8_t anim002_map[] = {
      |                                                               ^~~~~~~~~~~

**Comment by kisvegabor** [comment]
It seems the `#if LV_COLOR_DEPTH == 16 && LV_COLOR_16_SWAP != 0` is missing from the converted images. Have you removed them manually for some reason?

**Comment by ZhaoQiang-b45475** [comment]
The log shows no prototype for "lv_example_animimg", however, it is defined in "examples/widgets/lv_example_widgets.h". Is there anything I miss?
In addition, I ran "sudo apt-get install libpng-dev; cd tests; python ./build.py" locally, it passed.

examples/widgets/animimg/lv_example_animimg.c#L13
no previous prototype for ‘lv_example_animimg’ [-Werror=missing-prototypes]

File: examples/widgets/animimg/lv_example_animimg.c [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#include "../../../lvgl.h"
```
Comment (kisvegabor) [inline examples/widgets/animimg/lv_example_animimg.c]:
```suggestion
#include "../../lv_examples.h"
```

It should solve teh build issue. 

For consistency please rename `LV_USE_ANIM_IMG` to `LV_USE_ANIMIMG`.

**Review by kisvegabor — CHANGES_REQUESTED** [review]

File: examples/widgets/animimg/lv_example_animimg.c (hunk 1 above)
Comment (ZhaoQiang-b45475) [reply examples/widgets/animimg/lv_example_animimg.c]: Yes, it works, but I notice that the other examples (such as img, calendar) using "include ../../../lvgl.h".

Comment (kisvegabor) [reply examples/widgets/animimg/lv_example_animimg.c]: The includes in other examples were updated in master. 
See https://github.com/lvgl/lvgl/blob/master/examples/widgets/arc/lv_example_arc_1.c#L1

**Comment by ZhaoQiang-b45475** [comment]
Hi Kisvegabor,
Looks like there are no more comments, would you take any action on this PR?

**Comment by kisvegabor** [comment]
@turoksama 
This PR perfect example of what kind of external components I have imagined. You have mentioned that you are interested in helping to review them.  Do you have time to test it?

**Comment by turoksama** [comment]
@kisvegabor 
Hi there,sorry for the delay,I'll test it tomorrow.

**Review by turoksama — CHANGES_REQUESTED** [review]
Nice! The following needs to be updated:
1.
```static void lv_animimg_constructor(lv_obj_t * obj)  ```
is  incompatible pointer type with 
```void (*constructor_cb)(const struct _lv_obj_class_t * class_p, struct _lv_obj_t * obj);```
2.update lv_conf_template sync.

**Comment by ZhaoQiang-b45475** [comment]
@turoksama
Have updated, could you help to review again?

**Comment by turoksama** [comment]
@ZhaoQiang-b45475 
Thanks for your updates,  I'll test it in my hardware setup very soon.

**Review by turoksama — APPROVED** [review]
@kisvegabor 
Hi, there.
I can not get my simulator & hardware BOTH running base on latest v8 dev, looks like something bad happened when initialized, display driver data is broken(not the initialed one), did I miss something?

**Comment by embeddedt** [comment]
@turoksama Display and input drivers need to be `static` as of v8 (i.e. not go out of scope after registration). If not they will generally get corrupted.

**Comment by turoksama** [comment]
@embeddedt 
cool! got it.

**Comment by ZhaoQiang-b45475** [comment]
@turoksama
Any update?

**Comment by turoksama** [comment]
@ZhaoQiang-b45475 
Hi, I reviewed and passed days ago. No more recommendations. 
@kisvegabor 
It seems fine to me, what do you think?

**Comment by kisvegabor** [comment]
Cool, I merge it.

Thank you @ZhaoQiang-b45475 and @turoksama.

**Comment by kisvegabor** [comment]
I've just fixed a flickering after the last image and followed and LVGL API changes: 3b7bbddb

**Pull request merged**